src/flint/types/fmpz_mod_poly.pyx

```diff
@@ -0,0 +1,1297 @@
+            if not (<fmpz_mod_poly>left).ctx == (<fmpz_mod_poly>right).ctx:
+                raise ValueError("moduli must match")
+
+        # Case when right is not fmpz_mod_poly, try to convert to fmpz
+        elif typecheck(left, fmpz_mod_poly):
+            right = (<fmpz_mod_poly>left).ctx.any_as_fmpz_mod_poly(right)
+            if right is NotImplemented:
+                return NotImplemented
+
+        # Case when left is not fmpz_mod_poly, try to convert to fmpz
+        else:
+            left = (<fmpz_mod_poly>right).ctx.any_as_fmpz_mod_poly(left)
+            if left is NotImplemented:
+                return NotImplemented
+
+        res = fmpz_mod_poly.__new__(fmpz_mod_poly)
+        res.ctx = (<fmpz_mod_poly>left).ctx
+        check = fmpz_mod_poly_divides(
+            res.val, (<fmpz_mod_poly>left).val, (<fmpz_mod_poly>right).val, res.ctx.mod.val
+        )
+        if check == 0:
+            raise ValueError(
+                f"{right} does not divide {left}"
+            )
+
+        return res
+
+    def __truediv__(s, t):
+        return fmpz_mod_poly._div_(s, t)
+
+    def __rtruediv__(s, t):
+        return fmpz_mod_poly._div_(t, s)
+
+    @staticmethod
+    def _floordiv_(left, right):
+        cdef fmpz_mod_poly res
+
+        # Case when left and right are already fmpz_mod_poly
+        if typecheck(left, fmpz_mod_poly) and typecheck(right, fmpz_mod_poly):
+            if not (<fmpz_mod_poly>left).ctx == (<fmpz_mod_poly>right).ctx:
+                raise ValueError("moduli must match")
+
+        # Case when right is not fmpz_mod_poly, try to convert to fmpz
+        elif typecheck(left, fmpz_mod_poly):
+            right = (<fmpz_mod_poly>left).ctx.any_as_fmpz_mod_poly(right)
+            if right is NotImplemented:
+                return NotImplemented
+
+        # Case when left is not fmpz_mod_poly, try to convert to fmpz
+        else:
+            left = (<fmpz_mod_poly>right).ctx.any_as_fmpz_mod_poly(left)
+            if left is NotImplemented:
+                return NotImplemented
+
+        if not right.leading_coefficient().is_unit():
+            raise ValueError(f"The leading term of {right} must be a unit modulo N")
+
+        res = fmpz_mod_poly.__new__(fmpz_mod_poly)
+        res.ctx = (<fmpz_mod_poly>left).ctx
+        fmpz_mod_poly_div(
+            res.val, (<fmpz_mod_poly>left).val, (<fmpz_mod_poly>right).val, res.ctx.mod.val
+        )
+        return res 
+
+    def __floordiv__(self, other):
+        return fmpz_mod_poly._floordiv_(self, other)
+
+    def __rfloordiv__(self, other):
+        return fmpz_mod_poly._floordiv_(other, self)
+
+    def __pow__(self, e):
+        cdef fmpz_mod_poly res
+        if e < 0:
+            raise ValueError("Exponent must be non-negative")
+
+        res = fmpz_mod_poly.__new__(fmpz_mod_poly)
+        res.ctx = self.ctx
+        fmpz_mod_poly_pow(
+            res.val, self.val, (<ulong>e), self.ctx.mod.val
```

Comment:
It should be checked that `e` is an integer before calling `__new__`.

Reply:
What's the best way to do this? Something like `if e != (<ulong>e)`?

Reply:
I think something like
```
cdef ulong e_ulong = e
```